### PR TITLE
Allow tracking of global/environment property reads

### PR DESCRIPTION
**Pull request by maneely**
Re-reverting this as a starting point for #3432 and #5015 

…#4461)" (#4663)"

This reverts commit faf5e5d75f737df486e9005c4024738fa723da7d.

### CODE REVIEW

**Comment by maneely** [comment]
This effectively adds back in the property tracking work done back in August '19 (with merge resolutions).  This was reverted because of a failed RPS run with (IMO) questionable results.

Is it possible to get an RPS run before/after this change goes in (if accepted)?

**Comment by cdmihai** [comment]
> Is it possible to get an RPS run before/after this change goes in (if accepted)?

Just wrote this a few days back: https://github.com/microsoft/msbuild/pull/4997#issuecomment-570375824
Maybe I should add a piece of documentation on the current preferred way to of getting perf results.

For the OrchardCore case, you can repeat your measurements on master, your PR without property tracking, your PR with property tracking.

**Comment by maneely** [comment]
Per @cdmihai, launched an experimental build to run VS perf tests with latest changes. It passed.

File: src/Build/Evaluation/Evaluator.cs [hunk 1]
```diff
@@ -1334,21 +1341,24 @@ private void EvaluatePropertyElement(ProjectPropertyElement propertyElement)
...
-                P predecessor = _data.GetProperty(propertyElement.Name);
-
-                P property = _data.SetProperty(propertyElement, evaluatedValue, predecessor);
+                P property = _data.SetProperty(propertyElement, evaluatedValue);
 
-                if (predecessor != null)
+                if (Traits.Instance.LogPropertyTracking == 0)
                 {
-                    LogPropertyReassignment(predecessor, property, propertyElement.Location.LocationString);
+                    P predecessor = _data.GetProperty(propertyElement.Name);
```
Comment (Forgind) [inline src/Build/Evaluation/Evaluator.cs]: Perhaps I'm confused, but should this not go above line 1344?

Comment (maneely) [reply src/Build/Evaluation/Evaluator.cs]: It should. Refactoring.

File: src/Build/BackEnd/Components/Logging/LoggingContext.cs [hunk 2]
```diff
@@ -231,7 +231,7 @@ internal void LogBuildEvent(BuildEventArgs buildEvent)
         /// <summary>
         /// Log an error based on an exception
         /// </summary>
-        /// <param name="exception">The exception wich is to be logged</param>
+        /// <param name="exception">The exception which is to be logged</param>
```
[the remark this answers is not in the document]
Comment (maneely) [reply src/Build/BackEnd/Components/Logging/LoggingContext.cs]: Changed with next update.

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs [hunk 3]
```diff
@@ -29,10 +29,12 @@ public sealed class BinaryLogger : ILogger
...
         // version 6:
-        //   -  Ids and parent ids for the evaluation locations
+        //   - Ids and parent ids for the evaluation locations
         // version 7:
-        //   -  Include ProjectStartedEventArgs.GlobalProperties
-        internal const int FileFormatVersion = 7;
+        //   - Include ProjectStartedEventArgs.GlobalProperties
+        // version 8:
+        //   - new record kinds: EnvironmentVariableRead, PropertyReassignment, UninitializedPropertyRead
+        internal const int FileFormatVersion = 8;
```
[the remark this answers is not in the document]
Comment (maneely) [reply src/Build/Logging/BinaryLogger/BinaryLogger.cs]: Fixed in latest commit.

File: src/Build.UnitTests/Evaluation/Evaluator_Tests.cs [hunk 4]
```diff
@@ -4514,7 +4514,7 @@ public void VerifyMSBuildLogsAMessageWhenLocalPropertyCannotOverrideValueOfGloba
         [Fact]
         public void VerifyPropertyTrackingLoggingDefault()
         {
-            // Having nothing defined should default to only logging reassignments
```
Comment (cdmihai) [inline src/Build.UnitTests/Evaluation/Evaluator_Tests.cs]: I'd make this test a Theory to indeed assert that reassignments get logged when it's turned on.

Comment (cdmihai) [reply src/Build.UnitTests/Evaluation/Evaluator_Tests.cs]: Oh, nevermind, next test does exactly this.

**Pull request merged**